wcfsetup/install/files/lib/system/bbcode/MessageParser.class.php

```diff
@@ -132,9 +132,15 @@ public function parse($message, $enableSmilies = true, $enableHtml = false, $ena
 	 * @return	string		text
 	 */
 	protected function parseSmilies($text, $enableHtml = false) {
-		foreach ($this->smilies as $code => $html) {
-			//$text = preg_replace('~(?<!&\w{2}|&\w{3}|&\w{4}|&\w{5}|&\w{6}|&#\d{2}|&#\d{3}|&#\d{4}|&#\d{5})'.preg_quote((!$enableHtml ? StringUtil::encodeHTML($code) : $code), '~').'(?![^<]*>)~', $html, $text);
-			$text = preg_replace('~(?<=^|\s|<li>)'.preg_quote((!$enableHtml ? StringUtil::encodeHTML($code) : $code), '~').'(?=$|\s|</li>'.(!$enableHtml ? '|<br />' : '').')~', $html, $text);
+		$codes = array_keys($this->smilies);
+		$pattern = '~(?<=^|\s|<li>)'.str_replace('@', '|', preg_quote((!$enableHtml ? StringUtil::encodeHTML(implode('@', $codes)) : implode('|', $codes)), '~')).'(?=$|\s|</li>'.(!$enableHtml ? '|<br />' : '').')~';
```

Comment:
Personally I'd prefer the Regex class. I find the closure pretty ugly for that job. But first: Let @dtdesign take a look at the basic idea behind this PR.